stream_alert/alert_processor/outputs/credentials/provider.py

```diff
@@ -0,0 +1,654 @@
+
+        # Encrypt the creds and push them to S3
+        if not credentials.is_encrypted():
+            credentials.encrypt(self._region, kms_key_alias)
+
+        encrypted_credentials = credentials.data()
+        if not encrypted_credentials:
+            return True
+
+        return send_creds_to_s3(self._region, self._bucket, s3_key, encrypted_credentials)
+
+    def get_s3_key(self, descriptor):
+        """Returns an appropriate S3 bucket key for credentials relevant to this Output.
+
+        Args:
+            descriptor (str):
+
+        Returns:
+            string
+        """
+        return get_formatted_output_credentials_name(self._service_name, descriptor)
+
+    def get_s3_secrets_bucket(self):
+        """Returns an appropriate S3 bucket for all credentials relevant to this driver.
+
+        Returns:
+            string
+        """
+        return '{}.streamalert.secrets'.format(self._prefix)
+
+
+class LocalFileDriver(CredentialsProvidingDriver, FileDescriptorProvider, CredentialsCachingDriver):
+    """Driver for fetching credentials that are saved locally on the filesystem."""
+
+    def __init__(self, region, service_name):
+        self._region = region
+        self._service_name = service_name
+        self._temp_dir = self.get_local_credentials_temp_dir()
+
+    def load_credentials(self, descriptor):
+        local_cred_location = self.get_file_path(descriptor)
+        with open(local_cred_location, 'rb') as cred_file:
+            encrypted_credentials = cred_file.read()
+
+        return Credentials(encrypted_credentials, True, self._region)
+
+    def has_credentials(self, descriptor):
+        return os.path.exists(self.get_file_path(descriptor))
+
+    def save_credentials(self, descriptor, credentials):
+        if not credentials.is_encrypted():
+            LOGGER.error('Error: Writing unencrypted credentials to disk is disallowed.')
+            return False
+
+        with self.offer_fileobj(descriptor) as file_handle:
+            file_handle.write(credentials.data())
+        return True
+
+    @staticmethod
+    def clear():
+        """Removes the local secrets directory that may be left from previous runs"""
+        secrets_dirtemp_dir = LocalFileDriver.get_local_credentials_temp_dir()
+
+        # Check if the folder exists, and remove it if it does
+        if os.path.isdir(secrets_dirtemp_dir):
+            shutil.rmtree(secrets_dirtemp_dir)
+
+    def offer_fileobj(self, descriptor):
+        """Opens a file-like object and returns it.
+
+        If you use the return value in a `with` statement block then the file descriptor
+        will auto-close.
+
+        Return:
+            file object
+        """
+        file_path = self.get_file_path(descriptor)
+        if not os.path.exists(file_path):
+            os.makedirs(os.path.dirname(file_path))
+
+        return open(file_path, 'a+b')  # read+write and in binary mode
```

Comment:
can you point to when we want/need to have `write` perms for the files?

Reply:
In the S3Driver, the `offer_fileobj` is offered to the AWS Boto client in the `download_fileobj` API call, which writes the result into a file-like object.

Reply:
got it!